File: src/org/jgll/traversal/ModelBuilderVisitor.java
package org.jgll.traversal;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

import org.jgll.grammar.Character;
import org.jgll.grammar.CharacterClass;
import org.jgll.grammar.HeadGrammarSlot;
import org.jgll.grammar.TerminalGrammarSlot;
import org.jgll.grammar.slot.BodyGrammarSlot;
import org.jgll.grammar.slot.LastGrammarSlot;
import org.jgll.sppf.IntermediateNode;
import org.jgll.sppf.ListSymbolNode;
import org.jgll.sppf.NonterminalSymbolNode;
import org.jgll.sppf.PackedNode;
import org.jgll.sppf.SPPFNode;
import org.jgll.sppf.TerminalSymbolNode;
import org.jgll.util.Input;

import static org.jgll.traversal.SPPFVisitorUtil.*;

/**
 * ModelBuilderVisitor builds a data model by visiting an SPPF and 
 * calling appropriate models to the provided {@link NodeListener} 
 * interface.
 * 
 * The meta-data needed for building models, stored in grammar slots,
 * is automatically retrieved and passed to the given node listener.
 * 
 * @author Ali Afroozeh
 * 
 * @see NodeListener
 *
 */

@SuppressWarnings("unchecked")
public class ModelBuilderVisitor<T, U> implements SPPFVisitor {
	
	private NodeListener<T, U> listener;
	private Input input;
	
	public ModelBuilderVisitor(Input input, NodeListener<T, U> listener) {
		this.input = input;
		this.listener = listener;
	}

	@Override
	public void visit(TerminalSymbolNode terminal) {
		if(!terminal.isVisited()) {
			terminal.setVisited(true);
			if(terminal.getMatchedChar() == TerminalSymbolNode.EPSILON) {
				terminal.setObject(Result.skip());
			} else {
				Result<U> result = listener.terminal(terminal.getMatchedChar(), input.getPositionInfo(terminal.getLeftExtent(), terminal.getRightExtent()));
				terminal.setObject(result);
			}
		}		
	}

	@Override
	public void visit(NonterminalSymbolNode nonterminalSymbolNode) {
		removeIntermediateNode(nonterminalSymbolNode);
		
		if(!nonterminalSymbolNode.isVisited()) {
		
			nonterminalSymbolNode.setVisited(true);
			
			if(nonterminalSymbolNode.isAmbiguous()) {
				
				buildAmbiguityNode(nonterminalSymbolNode);
				
			// Lazy creation of the children of a keyword.
			} else if(nonterminalSymbolNode.isKeywordNode()) {
				HeadGrammarSlot head = (HeadGrammarSlot) nonterminalSymbolNode.getGrammarSlot();
				BodyGrammarSlot currentSlot = head.getAlternateAt(0).getFirstSlot();
				
				List<U> list = new ArrayList<>();
				int i = nonterminalSymbolNode.getLeftExtent();
				while(!(currentSlot instanceof LastGrammarSlot)) {
					CharacterClass terminal = (CharacterClass) ((TerminalGrammarSlot)currentSlot).getTerminal();
					assert terminal.getRanges().size() == 1;
					Result<U> result = listener.terminal(terminal.getRanges().get(0).getStart(), input.getPositionInfo(i, ++i));
					list.add(result.getObject());
					currentSlot = currentSlot.next();
				}
				
				LastGrammarSlot slot = (LastGrammarSlot) nonterminalSymbolNode.getFirstPackedNodeGrammarSlot();
				listener.startNode((T) slot.getObject());
				Result<U> result = listener.endNode((T) slot.getObject(), list, 
						input.getPositionInfo(nonterminalSymbolNode.getLeftExtent(), nonterminalSymbolNode.getRightExtent()));
				nonterminalSymbolNode.setObject(result);
			} else {
				LastGrammarSlot slot = (LastGrammarSlot) nonterminalSymbolNode.getFirstPackedNodeGrammarSlot();
				listener.startNode((T) slot.getObject());
				visitChildren(nonterminalSymbolNode, this);
				Result<U> result = listener.endNode((T) slot.getObject(), getChildrenValues(nonterminalSymbolNode), 
						input.getPositionInfo(nonterminalSymbolNode.getLeftExtent(), nonterminalSymbolNode.getRightExtent()));
				nonterminalSymbolNode.setObject(result);
			}
		}
	}

	private void buildAmbiguityNode(NonterminalSymbolNode nonterminalSymbolNode) {
		int nPackedNodes = 0;
		
		for(SPPFNode child : nonterminalSymbolNode.getChildren()) {
			PackedNode packedNode = (PackedNode) child;
			LastGrammarSlot slot = (LastGrammarSlot) packedNode.getGrammarSlot();
			listener.startNode((T) slot.getObject());
			packedNode.accept(this);
			Result<U> result = listener.endNode((T) slot.getObject(), getChildrenValues(packedNode), 
					input.getPositionInfo(packedNode.getLeftExtent(), packedNode.getRightExtent()));
			packedNode.setObject(result);
			if(result != Result.filter()) {
				nPackedNodes++;
			}
		}
		
		if(nPackedNodes > 1) {
			Result<U> result = listener.buildAmbiguityNode(getChildrenValues(nonterminalSymbolNode), 
					input.getPositionInfo(nonterminalSymbolNode.getLeftExtent(), nonterminalSymbolNode.getRightExtent()));
			nonterminalSymbolNode.setObject(result);
		}
	}

	@Override
	public void visit(IntermediateNode node) {
		// Intermediate nodes should be removed when visiting their parents.
		throw new RuntimeException("Should not be here!");
	}

	@Override
	public void visit(PackedNode packedNode) {
		removeIntermediateNode(packedNode);
		removeListSymbolNode(packedNode);
		if(!packedNode.isVisited()) {
			packedNode.setVisited(true);
			visitChildren(packedNode, this);
		}
	}

	@Override
	public void visit(ListSymbolNode listNode) {
		removeListSymbolNode(listNode);
		if(!listNode.isVisited()) {
			listNode.setVisited(true);
			if(listNode.isAmbiguous()) {
				buildAmbiguityNode(listNode);
			} else {
				LastGrammarSlot slot = (LastGrammarSlot) listNode.getFirstPackedNodeGrammarSlot();
				listener.startNode((T) slot.getObject());
				visitChildren(listNode, this);
				Result<U> result = listener.endNode((T) slot.getObject(), getChildrenValues(listNode), 
						input.getPositionInfo(listNode.getLeftExtent(), listNode.getRightExtent()));
				listNode.setObject(result);
			}
		}
	}
	
	private Iterable<U> getChildrenValues(final SPPFNode node) {

		final Iterator<SPPFNode> iterator = node.getChildren().iterator();

		return new Iterable<U>() {

			@Override
			public Iterator<U> iterator() {
				return new Iterator<U>() {

					private SPPFNode next;

					@Override
					public boolean hasNext() {
						while (iterator.hasNext()) {
							next = iterator.next();
							
							if(next.getObject() == Result.filter()) {
								node.setObject(Result.filter());
								// Go to the next child
								if(!iterator.hasNext()) {
									return false;
								}
								next = iterator.next();
							} 
							
							else if(next.getObject() == Result.skip()) {
								// Go to the next child
								if(!iterator.hasNext()) {
									return false;
								}
								next = iterator.next();
							}
							
							else if(next.getObject() == null) {
								if(!iterator.hasNext()) {
									return false;
								}
								next = iterator.next();
							}
							
							else {
								return true;								
							}
						}
						return false;
					}

					@Override
					public U next() {
						return ((Result<U>) next.getObject()).getObject();
					}

					@Override
					public void remove() {
						throw new UnsupportedOperationException();
					}
				};
				
			}
		};
	}

	
}

File: src-scala/org/jgll_staged/traversal/ModelBuilderVisitor.scala
package org.jgll_staged.traversal

import java.util.ArrayList
import java.util.Iterator
import java.util.List
import org.jgll_staged.grammar.java.lang.Character
import org.jgll_staged.grammar.CharacterClass
import org.jgll_staged.grammar.slot.BodyGrammarSlot
import org.jgll_staged.grammar.slot.HeadGrammarSlot
import org.jgll_staged.grammar.slot.LastGrammarSlot
import org.jgll_staged.grammar.slot.TerminalGrammarSlot
import org.jgll_staged.sppf.IntermediateNode
import org.jgll_staged.sppf.ListSymbolNode
import org.jgll_staged.sppf.NonterminalSymbolNode
import org.jgll_staged.sppf.PackedNode
import org.jgll_staged.sppf.SPPFNode
import org.jgll_staged.sppf.TerminalSymbolNode
import org.jgll_staged.util.Input
import org.jgll_staged.traversal.SPPFVisitorUtil._
//remove if not needed
import scala.collection.JavaConversions._

class ModelBuilderVisitor[T, U](private var input: Input, private var listener: NodeListener[T, U])
    extends SPPFVisitor {

  override def visit(terminal: TerminalSymbolNode) {
    if (!terminal.isVisited) {
      terminal.setVisited(true)
      if (terminal.getMatchedChar == TerminalSymbolNode.EPSILON) {
        terminal.setObject(Result.skip())
      } else {
        val result = listener.terminal(terminal.getMatchedChar, input.getPositionInfo(terminal.getLeftExtent, 
          terminal.getRightExtent))
        terminal.setObject(result)
      }
    }
  }

  override def visit(nonterminalSymbolNode: NonterminalSymbolNode) {
    removeIntermediateNode(nonterminalSymbolNode)
    if (!nonterminalSymbolNode.isVisited) {
      nonterminalSymbolNode.setVisited(true)
      if (nonterminalSymbolNode.isAmbiguous) {
        buildAmbiguityNode(nonterminalSymbolNode)
      } else if (nonterminalSymbolNode.isKeywordNode) {
        val head = nonterminalSymbolNode.getGrammarSlot.asInstanceOf[HeadGrammarSlot]
        var currentSlot = head.getAlternateAt(0).getFirstSlot
        val list = new ArrayList[U]()
        val i = nonterminalSymbolNode.getLeftExtent
        while (!(currentSlot.isInstanceOf[LastGrammarSlot])) {
          val terminal = currentSlot.asInstanceOf[TerminalGrammarSlot].getTerminal.asInstanceOf[CharacterClass]
          assert(terminal.getRanges.size == 1)
          val result = listener.terminal(terminal.getRanges.get(0).getStart, input.getPositionInfo(i, 
            i))
          list.add(result.getObject)
          currentSlot = currentSlot.next()
        }
        val slot = nonterminalSymbolNode.getFirstPackedNodeGrammarSlot.asInstanceOf[LastGrammarSlot]
        listener.startNode(slot.getObject.asInstanceOf[T])
        val result = listener.endNode(slot.getObject.asInstanceOf[T], list, input.getPositionInfo(nonterminalSymbolNode.getLeftExtent, 
          nonterminalSymbolNode.getRightExtent))
        nonterminalSymbolNode.setObject(result)
      } else {
        val slot = nonterminalSymbolNode.getFirstPackedNodeGrammarSlot.asInstanceOf[LastGrammarSlot]
        listener.startNode(slot.getObject.asInstanceOf[T])
        visitChildren(nonterminalSymbolNode, this)
        val result = listener.endNode(slot.getObject.asInstanceOf[T], getChildrenValues(nonterminalSymbolNode), 
          input.getPositionInfo(nonterminalSymbolNode.getLeftExtent, nonterminalSymbolNode.getRightExtent))
        nonterminalSymbolNode.setObject(result)
      }
    }
  }

  private def buildAmbiguityNode(nonterminalSymbolNode: NonterminalSymbolNode) {
    var nPackedNodes = 0
    for (child <- nonterminalSymbolNode.getChildren) {
      val packedNode = child.asInstanceOf[PackedNode]
      val slot = packedNode.getGrammarSlot.asInstanceOf[LastGrammarSlot]
      listener.startNode(slot.getObject.asInstanceOf[T])
      packedNode.accept(this)
      val result = listener.endNode(slot.getObject.asInstanceOf[T], getChildrenValues(packedNode), input.getPositionInfo(packedNode.getLeftExtent, 
        packedNode.getRightExtent))
      packedNode.setObject(result)
      if (result != Result.filter()) {
        nPackedNodes += 1
      }
    }
    if (nPackedNodes > 1) {
      val result = listener.buildAmbiguityNode(getChildrenValues(nonterminalSymbolNode), input.getPositionInfo(nonterminalSymbolNode.getLeftExtent, 
        nonterminalSymbolNode.getRightExtent))
      nonterminalSymbolNode.setObject(result)
    }
  }

  override def visit(node: IntermediateNode) {
    throw new RuntimeException("Should not be here!")
  }

  override def visit(packedNode: PackedNode) {
    removeIntermediateNode(packedNode)
    removeListSymbolNode(packedNode)
    if (!packedNode.isVisited) {
      packedNode.setVisited(true)
      visitChildren(packedNode, this)
    }
  }

  override def visit(listNode: ListSymbolNode) {
    removeListSymbolNode(listNode)
    if (!listNode.isVisited) {
      listNode.setVisited(true)
      if (listNode.isAmbiguous) {
        buildAmbiguityNode(listNode)
      } else {
        val slot = listNode.getFirstPackedNodeGrammarSlot.asInstanceOf[LastGrammarSlot]
        listener.startNode(slot.getObject.asInstanceOf[T])
        visitChildren(listNode, this)
        val result = listener.endNode(slot.getObject.asInstanceOf[T], getChildrenValues(listNode), input.getPositionInfo(listNode.getLeftExtent, 
          listNode.getRightExtent))
        listNode.setObject(result)
      }
    }
  }

  private def getChildrenValues(node: SPPFNode): java.lang.Iterable[U] = {
    val iterator = node.getChildren.iterator()
    new java.lang.Iterable[U]() {

      override def iterator(): Iterator[U] = {
        new Iterator[U]() {

          private var next: SPPFNode = _

          override def hasNext(): Boolean = {
            while (iterator.hasNext) {
              next = iterator.next()
              if (next.getObject == Result.filter()) {
                node.setObject(Result.filter())
                if (!iterator.hasNext) {
                  return false
                }
                next = iterator.next()
              } else if (next.getObject == Result.skip()) {
                if (!iterator.hasNext) {
                  return false
                }
                next = iterator.next()
              } else if (next.getObject == null) {
                if (!iterator.hasNext) {
                  return false
                }
                next = iterator.next()
              } else {
                return true
              }
            }
            false
          }

          override def next(): U = {
            next.getObject.asInstanceOf[Result[U]].getObject
          }

          override def remove() {
            throw new UnsupportedOperationException()
          }
        }
      }
    }
  }
}